Adding regional recursive filter functionality

Masanori-NOAA opened the pull request:
This PR is to add regional recursive filter functionality to GSIbec. The branch was compiled both in GDASApp and RDASApp on Orion.

Masanori-NOAA committed:
Add regional RF

Masanori-NOAA committed:
Modifications to include regional RF

rtodling commented on src/gsibec/gsi/CMakeLists.txt:
```diff
@@ -31,13 +31,13 @@ general_specmod.f90
 general_sub2grid_mod.f90
 general_transform.f90
-genqsat.F90
+genqsat.f90
```
Pleaes let's not do this: please leave the files as F90 as they were even if no directives are in the file; it is extremely inconvenient to have to turn these files into F90 when debugging using directives. 

This applies to a number of cases ...

Masanori-NOAA replied:
Thank you for your review. I’ve addressed your comments and pushed the changes to the branch.

rtodling replied:
good

rtodling reviewed (commented):
I am pulling this into my work env to give it a try - make sure I can reproduce what I have without these changes.

rtodling commented on src/gsibec/gsi/m_berror_stats.f90:
```diff
@@ -103,15 +105,18 @@ module m_berror_stats
    if(.not.allocated(varq)) then
      if (getindex(cvars3d,'q')>0) then
-        allocate(varq(mlat,msig))
+        !allocate(varq(mlat,msig))
```
what happened here? What is the line commented and uncommented - this is the original code w/ now an added commented line.

rtodling commented on src/gsibec/gsi/m_berror_stats.f90:
```diff
@@ -103,15 +105,18 @@ module m_berror_stats
    endif
    if(.not.allocated(varcw)) then
-      allocate(varcw(mlat,msig))
+      !allocate(varcw(mlat,msig))
```
same as above: what happened here? What is the line commented and uncommented - this is the original code w/ now an added commented line.

Masanori-NOAA committed:
Updated based on PR feedback

Masanori-NOAA committed:
Delete the commented line

Masanori-NOAA commented:
Thank you for your comment. I apologize for the lack of explanation. 
In the GSI code, I confirmed that the second dimension of the arrays “varq” and “varcw” is not msig (the number of vertical levels for background error), but nsig (the number of sigma levels). In the global analysis, nsig and msig are same so I think there is no problem in the original code for global analysis. But in the regional analysis, they differ. Therefore, I modified the second dimension to nsig in order to follow the GSI code. Also, I forgot to remove the commented-out line during the processs. I deleted the commented-out line. Thank you for pointing that out.

rtodling reviewed (approved):
These are good by me now. Thank you.
